sync/awsdatabricks.py

```diff
@@ -262,6 +248,73 @@ def _get_cluster_report(
     )
 
 
+def _get_cluster_instances(cluster: dict) -> dict | DatabricksError:
+    cluster_instances = None
+    aws_region_name = DB_CONFIG.aws_region_name
+
+    cluster_id = cluster["cluster_id"]
+    cluster_log_dest = _cluster_log_destination(cluster)
+
+    if cluster_log_dest:
+        s3 = boto.client("s3")
+        (_, bucket, base_prefix) = cluster_log_dest
+        cluster_instances_file_key = f"{base_prefix}/sync_data/cluster_instances.json"
+
+        try:
+            cluster_instances_file_response = s3.get_object(
+                Bucket=bucket, Key=cluster_instances_file_key
+            )
+            cluster_instances = orjson.loads(cluster_instances_file_response["Body"].read())
+        except ClientError as ex:
+            if ex.response["Error"]["Code"] == "NoSuchKey":
+                logger.warning(
+                    f"Could not find sync_data/cluster_instances.json for cluster: {cluster_id}"
+                )
+            else:
+                logger.error(
+                    f"Unexpected error encountered while attempting to fetch sync_data/cluster_instances.json: {ex}"
+                )
+
+    # If this cluster does not have the "Sync agent" configured, attempt a best-effort snapshot of the instances that
+    #  are associated with this cluster
+    if cluster_instances is None:
```

Comment:
That's true, though I can't see `cluster_instances` ever being simply `{}`. The reason is because `ec2.describe_instances` will always return a non-empty dict, even if no matching instances are found. The response looks something like -

```python
{
  "Reservations": [],
  "HTTPResponseMetadata": {...}  // This key name be wrong
}
```

And since we are preserving that "shape" during the polling as well, there will always be at least the `Reservations` key in the dict, if it exists.

This is a long way of saying that `if cluster_instances is None` is not really checking for "did we find instances?", but rather for "were we polling during the job execution?"